### DISCUSSION TITLE
Survey วิธีการออกแบบเกม

### DISCUSSION THREAD

**Discussion by ppbasleng** [Logical-Learning-Game/Document#29] [Show and tell]
Survey เอกสารและงานที่เกี่ยวกับการสร้างเกม

**Comment by ppbasleng** [comment]
## Dynamic Game Difficulty Balancing
https://en.wikipedia.org/wiki/Dynamic_game_difficulty_balancing

## 

เป็นกระบวนการปรับเปลี่ยนค่า parameter ต่างๆภายในเกมแบบ real time โดยจะขึ้นอยู่กับความสามารถของผู้เล่น เหตุผลหลักก็เพื่อทำให้ผู้เล่นไม่เบื่อไปกับเกมที่ง่ายเกินไป หรือท้อใจกับเกมที่ยากเกินไป เป้าหมายคือทำบาลานซ์ให้ผู้เล่นยังรู้สึกพอใจความระดับความยากง่ายของเกมตั้งแต่เริ่มต้นจบถึงเกมจบ

ตัว Parameter ที่มักจะปรับเปลี่ยน
- ความเร็วศัตรู
- พลังชีวิตของศัตรู
- จำนวนของศัตรู
- จำนวนของการเพิ่มพลังผู้เล่น
- พลังของผู้เล่น
- พลังของศัตรู
- ระยะเวลาของการเล่น

## ความเห็นส่วนตัว
จากการศึกษาและทดลองเล่นด้วยตัวเองนั้น จะมีบางเกมที่ใช้ AI ในการคำนวณเรื่อง Parameter ที่ต้องปรับด้วย
แต่ในประเภทของเกมที่มักจะมีการใส่ระบบดังกล่าวมักจะเป็นเกมประเภท FPS/TPS (เกมยิงมุมมองบุคคลต่างๆ) หรือเกมประเภท mmorpg/action rpg (เกมต่อสู้สวมบทบาทแบบแอคชั่น/ออนไลน์) เพียงเท่านั้น และเห็นเกมประเภท Puzzle ได้ค่อนข้างน้อย
เกมตัวอย่างที่มีการนำวิธีการดังกล่าวไปใช้ในการปรับค่า คือ 
**Zanac(1986)** เกมประเภท Shoot 'em up ที่มีการนำ adaptive AI มาใช้ในการปรับระดับความยากของเกมโดยจะดูจากระดับการเล่นของผู้เล่น อัตราการยิง และระดับความสามารถในการป้องกันของยานผู้เล่น

![Image](https://user-images.githubusercontent.com/54877995/185538673-76f9d961-ff02-4597-8efb-8062a81ea14d.png)

**Resident Evil4 (2005)**  TPS เกมยิงมุมมองบุคคลที่สาม ที่จากในหนังสือ [resident evil 4 official strategy guide](https://www.amazon.com/Resident-Official-Strategy-Bradygames-Signature/dp/0744003466) ได้เผยว่าในเกมมีการใช้ระบบ Difficulty Scale ที่มีตั้งแต่ค่า 1 - 10 ในการแอบปรับระดับความยากของเกมให้กับผู้เล่น เพื่อเพิ่มระดับความยากในการเล่น เช่น พฤติกรรมของซอมบี้ศัตรู ความถึกทนของศัตรู โดยระดับจะขึ้นอยู่กับความสามารถในการเล่นของผู้เล่น เช่น อัตราการตาย หรืออัตราการยิงติดคริติคอล (ยิงหัว)

![Image](https://user-images.githubusercontent.com/54877995/185539274-03b73a26-0ef3-40df-8c69-29fbec198333.png)

**Fishdom (2008)** เกมประเภท Match-3 คล้ายๆ candy crush ที่มีการปรับ time limit ในการผ่านด่านโดยดูว่าผู้เล่นเล่นได้ดีแค่ไหน (ไม่เจอเอกสารอ้างอิงที่เกี่ยวข้อง)

![Image](https://user-images.githubusercontent.com/54877995/185540402-23e649da-34d5-4843-bd22-3c5115fc0570.png)

## เอกสารที่น่าจะเกี่ยวข้อง
Dynamic Difficulty Adjustment 
[DDA_paper.pdf](https://github.com/Logical-Learning-Game/Document/files/9379021/DDA_paper.pdf)

**Comment by ppbasleng** [comment]
## การศึกษาองค์ประกอบของการออกแบบเกม Puzzle
https://www.youtube.com/watch?v=zsjC6fa_YBg&ab_channel=GameMaker%27sToolkit
##
* กฎเหล็กของเกม (Mechanics)
  * หลักการของเกมที่ใช้ควบคุมว่าเกมแต่ละเกมทำงานอย่างไร
  * สิ่งที่ผู้เล่นทำในเกมได้และข้อจำกัด จะทำให้เกิดเกม Puzzle
  * ความฉลาดของกฏมีผลโดยตรงกับจำนวนความหลากหลายและระดับความยากของเกม
  * อุปกรณ์เสริม (temporary tools) สามารถใช้ในการเป็นส่วนเสริมให้กับ mechanic หลักได้
    *  บางเกมก็ใช้อุปกรณ์เสริมเป็น gimmick ของเกมไปเลย
  * เกม Puzzle จำเป็นต้องมี **เป้าหมาย (Goal)** ซึ่งก็คือ สิ่งที่ผู้เล่นต้องบรรลุในด่าน ยกตัวอย่างเป้าหมายเช่น ทางออกจากด่าน หรือ ของที่ต้องเก็บในด่าน
* อุปสรรค/ความขัดแย้ง (The Catch)
   * เกม puzzle ที่ดี จะต้องเป็นเกมที่เล่นเกี่ยวกับอุปสรรคที่ผู้เล่นจะเจอในด่าน
   * ตัวอย่างของการอธิบายความหมายของอุปสรรค คือ ในห้องๆหนึ่ง มีประตูที่ปิดอยู่ และมีแท่นเหยียบ หากยืนบนแท่น ประตูจะเปิด แต่เมื่อเดินไปที่ประตู ก็จะไม่ได้เหยียบแท่น ประตูก็จะปิด วิธีคือต้องหาของมาวางบนแท่นเพื่อให้ประตูเปิด เพื่อนที่จะออกจากห้อง
* การรับรู้ การค้นพบ(The Revelation)
  * การเล่นเกมประเภท Puzzle ก็เปรียบเหมือนการรับรู้ หรือ ค้นพบสิ่งใหม่ๆ 
  * ความสนุกของการเล่นเกม Puzzle ก็คือการที่เราไขปริศนาจนได้ แบบ "อ๋อ รู้แล้ววว" 
  * การออกแบบPuzzle ที่ดีไม่ควรทำให้ปัญหาที่ผู้เล่นเจอชัดเจนมาก และควรเปิดโอกาสให้ผู้เล่นได้ทดลองคิดนอกกรอบ หากผู้เล่นรู้คำตอบจากการดู walkthrough ก็จะต้องประมาณ "อ่าว เพิ่งรู้ว่าทำแบบนี้ก็ได้ด้วย"
* การหลอกล่อด้วยข้อสรุปลวง (The Assumption)
  * ในบางครั้งที่เกมเริ่มด่านมา เมื่อเราวิเคราะห์จากสภาพแวดล้อมแล้วว่าต้องแก้ปัญหายังไง บางครั้งวิธีแก้นั้นจะไม่ได้ถูกเสมอไป
  * เป็นการสร้างความรู้สึก หลงทาง (strayed) ให้กับผู้เล่น
  * การออกแบบเกมโดยใช้วิธีนี้จะหมายความว่ามีโอกาสสูงมากที่ผู้เล่นจะเล่นไม่ผ่านในครั้งแรก
* การนำเสนอ (The Presentation)
  * เกมที่ดีควรที่ความ minimalist ใน Mechanic ของตัวมันเอง
  * เกมที่มีหลากหลาย element มากเกินไปจะทำให้ระบบเกมซับซ้อนเกินไป
  * เกมควรจะมี feedback ที่ชัดเจนว่าทำอะไร แล้วเกิดอะไรขึ้น
*  ระดับของเกม (The Curve)
   * Puzzle ในเกมส่วนใหญ่จะถูกสร้างอยู่บน Puzzle ก่อนที่เคยผ่านมาแล้ว
   * Puzzle ส่วนใหญ่จะเน้นให้ผู้เล่นใช้ทักษะที่รู้มาจากด่านก่อนๆหน้า(หรือที่เกมสอน) มาใช้ในการแก้ปัญหาในด่านถัดๆไป
   * การออกแบบเกม Puzzle ถ้าหากให้ผู้เล่นเล่นด่านแบบสุ่ม แทบไม่มีทางที่ผู้เล่นใหม่จะสามารถเล่นด่านผ่านได้เลย
   * เกม Puzzle ควรที่จะค่อยๆ เพิ่มระดับความยากของด่านขึ้นทีละเล็กละน้อย
   * ความยากง่ายของเกมขึ้นอยู่กับ 4 กฏเกณท์
      * จำนวนของวิธีแก้ปัญหา (ยิ่งน้อย ยิ่งยาก)
      * จำนวนของ step ที่ผู้เล่นกระทำ (ยิ่งเยอะ ยิ่งยาก แต่ยิ่งมากไปก็ไม่ดี)
      * จำนวนของทางเลือกที่ผู้เล่นต้องเลือกใช้/ตัดสินใจ
      * ความคุ้นชินกับ Mechanic ในเกม ที่ผู้เล่นเคยสัมผัสมาก่อนหน้า

 * **เกมประเภท Puzzle เป็นเกมที่จำเป็นต้องมีการ "ทดลองเล่น" มากที่สุดนับจากทุกแนวเกม**
## สรุปทั้งหมด
เกม Puzzle ที่ดี ควรเป็นเกมที่มี Mechanic และกฏเกมที่ชัดเจน และมีอุปสรรคที่ทำให้ผู้เล่นรู้สึกว่ายากที่จะแก้เมื่อเห็นในครั้งแรก
เกมที่ดีจะพยายามทำให้ผู้เล่นคิดกว้างๆ และค้นพบวิธีแก้ที่ซ่อนอยู่จากกฏของเกม

หากเกม Puzzle รู้สึกขาดๆ อาจจะหมายความว่า
- เกมนั้นยังไม่มีทิศทางที่ชัดเจน (Key aspect)
- เกมง่ายเกินไป
- เกมไม่มีตัวหลอกล่อ (assumption)
- เกมขาดการค้นพบสิ่งใหม่ๆ (revelation) เกมก็จะกลายเป็นเหมือนการจัดการปัญหายุ่งๆ อย่างเดียว
**เกม Puzzle เป็นแนวเกมที่ต้องใช้เวลามากที่สุดในการพัฒนา Mechanic ของเกม**

**Comment by Chuckboliver** [comment]
คุยวันนี้ครับ (17/08/2022)

![Image](https://user-images.githubusercontent.com/43881371/185058633-903f0ee2-5cf8-476e-8767-3ad950f26015.png)

นำ Puzzle จาก blockly game มาปรับให้เข้ากับเกม

รื้อ Mechanic ของเกมในการควบคุมตัวละครอื่นที่ไม่ใช่ตัวละครหลักมาใช้ในการผ่านด่าน
วางแผนลดการใช้ RL มาใช้ AI แทน (คิดว่า AI สามารถทำแทน RL ได้)

วางแผน Scope Project 1
Dev ฟังก์ชั่นพื้นฐานของเกมด้วย Unity (Grid base movement, tile)
ทดลอง Algorithm AI ด้วยภาษาอื่น (หรือ ภาษาเดียวกันกับ Unity Engine) แบบ Text-base
ทดลองการใช้เซิฟเวอร์กับ Unity
ออกแบบ Puzzle และ Mechanic ของเกม

วางแผน Scope Project 2
Dev ระบบโดยรวม (ทุกอย่าง)

**Comment by Chuckboliver** [comment]
Search base PCG

[2019_10_30-ProceduralContentGeneration_SearchGAsEtc.pdf](https://github.com/Logical-Learning-Game/Document/files/9357505/2019_10_30-ProceduralContentGeneration_SearchGAsEtc.pdf)

**Comment by Chuckboliver** [comment]
P NP NP-complete NP-had
https://kobkrit.com/%E0%B8%AA%E0%B8%A3%E0%B8%B8%E0%B8%9B-p-np-np-complete-np-hard-%E0%B9%81%E0%B8%A5%E0%B8%B0-p-np-%E0%B8%AB%E0%B8%A3%E0%B8%B7%E0%B8%AD%E0%B9%84%E0%B8%A1%E0%B9%88-%E0%B8%AE%E0%B9%88%E0%B8%B2%E0%B9%86-645186ff736d

**Comment by ppbasleng** [comment]
## Statistical Modelling of Level Difficulty in Puzzle Games
https://arxiv.org/pdf/2107.03305.pdf

#

**Comment by ppbasleng** [comment]
## A System for Procedurally Generating Puzzles for Games

[871036566-MIT.pdf](https://github.com/Logical-Learning-Game/Document/files/9379286/871036566-MIT.pdf)
